service/history/shard/context_test.go

```diff
@@ -365,3 +367,70 @@ func (s *contextSuite) TestDeleteWorkflowExecution_ErrorAndContinue_Success() {
 	s.NoError(err)
 	s.Equal(tasks.DeleteWorkflowExecutionStageCurrent|tasks.DeleteWorkflowExecutionStageMutableState|tasks.DeleteWorkflowExecutionStageVisibility|tasks.DeleteWorkflowExecutionStageHistory, stage)
 }
+
+func (s *contextSuite) TestAcquireShardOwnershipLostErrorIsNotRetried() {
+	s.mockShardManager.EXPECT().UpdateShard(gomock.Any(), gomock.Any()).
+		Return(&persistence.ShardOwnershipLostError{}).Times(1)
+	closed := make(chan struct{})
+	s.mockShard.closeCallback = func(impl *ContextImpl) {
+		close(closed)
+	}
+	s.mockShard.state = contextStateAcquiring
+	s.mockShard.acquireShardRetryPolicy = backoff.NewExponentialRetryPolicy(time.Nanosecond).
+		WithMaximumAttempts(5)
+
+	go s.mockShard.acquireShard()
+
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	case <-closed:
+		timer.Stop()
+	case <-timer.C:
+		s.Fail("shard should have been closed")
+	}
+}
+
+func (s *contextSuite) TestAcquireShardNonOwnershipLostErrorIsRetried() {
+	s.mockShardManager.EXPECT().UpdateShard(gomock.Any(), gomock.Any()).
+		Return(fmt.Errorf("temp error")).Times(6)
+	closed := make(chan struct{})
+	s.mockShard.closeCallback = func(impl *ContextImpl) {
+		close(closed)
+	}
+	s.mockShard.state = contextStateAcquiring
+	s.mockShard.acquireShardRetryPolicy = backoff.NewExponentialRetryPolicy(time.Nanosecond).
+		WithMaximumAttempts(5)
+
+	go s.mockShard.acquireShard()
+
+	timer := time.NewTimer(100 * time.Second)
+	select {
+	case <-closed:
+		timer.Stop()
+	case <-timer.C:
+		s.Fail("shard should have been closed")
```

Comment:
is this because the retry policy has max attempt of 5 and mock return error 6 times? Can we add one test case to verify that acquire shard succeed after a few errors and evenually succeed with retry?

Reply:
This is because we call the operation before checking the attempt count constraint here: https://github.com/temporalio/temporal/blob/2f84879686db552a9ca1bd6ea60fc099fa2557fa/common/backoff/retry.go#LL143

It looks like a bug AFAICT.

Reply:
I added the other test case